File: Week 20/1.Backtracking.java
/*
 * WEEK 20 - BACKTRACKING
 * Topic: Backtracking Template + Classic Problems
 *
 * BACKTRACKING:
 * Build a solution incrementally, step by step.
 * At each step, if the current partial solution cannot lead to a valid complete solution,
 * ABANDON it and backtrack to try other options.
 *
 * TEMPLATE:
 * void backtrack(State state) {
 *     if (isSolution(state)) { recordSolution(); return; }
 *     for (choice : getChoices(state)) {
 *         if (isValid(choice, state)) {
 *             makeChoice(choice, state);   // explore
 *             backtrack(state);
 *             undoChoice(choice, state);   // backtrack
 *         }
 *     }
 * }
 *
 * KEY: The "undo" step is what makes backtracking different from simple DFS.
 *
 * Time complexity: Usually exponential in the worst case (O(k^n) or O(n!)).
 * Pruning reduces the practical runtime significantly.
 *
 * PROBLEMS COVERED:
 * 1. Permutations of a list of numbers
 * 2. Subsets (power set)
 * 3. N-Queens
 * 4. Sudoku Solver
 * 5. Word Search in grid
 */

import java.util.*;

public class Backtracking {

    // PROBLEM 1: All Permutations
    // Time: O(n * n!), Space: O(n)
    static List<List<Integer>> permutations(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        permute(nums, 0, result);
        return result;
    }
    static void permute(int[] nums, int start, List<List<Integer>> result) {
        if (start == nums.length) {
            List<Integer> perm = new ArrayList<>();
            for (int x : nums) perm.add(x);
            result.add(perm);
            return;
        }
        for (int i = start; i < nums.length; i++) {
            swap(nums, start, i);           // choose
            permute(nums, start + 1, result); // explore
            swap(nums, start, i);           // unchoose (backtrack)
        }
    }
    static void swap(int[] nums, int i, int j) { int t = nums[i]; nums[i] = nums[j]; nums[j] = t; }

    // PROBLEM 2: All Subsets (Power Set)
    // Time: O(2^n * n), Space: O(n)
    static List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        subsetsHelper(nums, 0, new ArrayList<>(), result);
        return result;
    }
    static void subsetsHelper(int[] nums, int idx, List<Integer> current, List<List<Integer>> result) {
        result.add(new ArrayList<>(current)); // add current subset
        for (int i = idx; i < nums.length; i++) {
            current.add(nums[i]);           // include nums[i]
            subsetsHelper(nums, i + 1, current, result);
            current.remove(current.size() - 1); // exclude nums[i] (backtrack)
        }
    }

    // PROBLEM 3: N-Queens
    // Place N queens on NxN board such that no two queens attack each other.
    // Time: O(N!), Space: O(N)
    static List<List<String>> nQueens(int n) {
        List<List<String>> result = new ArrayList<>();
        int[] queens = new int[n]; // queens[row] = column of queen in that row
        Arrays.fill(queens, -1);
        nQueensHelper(queens, 0, n, result);
        return result;
    }
    static void nQueensHelper(int[] queens, int row, int n, List<List<String>> result) {
        if (row == n) { result.add(buildBoard(queens, n)); return; }
        for (int col = 0; col < n; col++) {
            if (isSafe(queens, row, col)) {
                queens[row] = col;
                nQueensHelper(queens, row + 1, n, result);
                queens[row] = -1; // backtrack
            }
        }
    }
    static boolean isSafe(int[] queens, int row, int col) {
        for (int r = 0; r < row; r++) {
            if (queens[r] == col) return false; // same column
            if (Math.abs(queens[r] - col) == Math.abs(r - row)) return false; // diagonal
        }
        return true;
    }
    static List<String> buildBoard(int[] queens, int n) {
        List<String> board = new ArrayList<>();
        for (int row = 0; row < n; row++) {
            char[] line = new char[n];
            Arrays.fill(line, '.');
            line[queens[row]] = 'Q';
            board.add(new String(line));
        }
        return board;
    }

    // PROBLEM 4: Word Search in Grid
    // Find if word exists in grid by traversing adjacent cells (up/down/left/right).
    static boolean wordSearch(char[][] board, String word) {
        int m = board.length, n = board[0].length;
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++)
                if (wordDFS(board, word, i, j, 0)) return true;
        return false;
    }
    static boolean wordDFS(char[][] board, String word, int i, int j, int k) {
        if (k == word.length()) return true;
        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length) return false;
        if (board[i][j] != word.charAt(k)) return false;
        char temp = board[i][j];
        board[i][j] = '#'; // mark as visited
        boolean found = wordDFS(board, word, i+1, j, k+1)
                     || wordDFS(board, word, i-1, j, k+1)
                     || wordDFS(board, word, i, j+1, k+1)
                     || wordDFS(board, word, i, j-1, k+1);
        board[i][j] = temp; // restore (backtrack)
        return found;
    }

    public static void main(String[] args) {
        // Permutations
        System.out.println("=== Permutations of [1,2,3] ===");
        List<List<Integer>> perms = permutations(new int[]{1, 2, 3});
        perms.forEach(System.out::println);

        // Subsets
        System.out.println("\n=== Subsets of [1,2,3] ===");
        subsets(new int[]{1, 2, 3}).forEach(System.out::println);

        // N-Queens
        System.out.println("\n=== 4-Queens Solutions ===");
        List<List<String>> solutions = nQueens(4);
        System.out.println("Number of solutions: " + solutions.size());
        for (List<String> sol : solutions) {
            sol.forEach(System.out::println);
            System.out.println("---");
        }

        // Word Search
        System.out.println("=== Word Search ===");
        char[][] grid = {
            {'A','B','C','E'},
            {'S','F','C','S'},
            {'A','D','E','E'}
        };
        System.out.println("Search 'ABCCED': " + wordSearch(grid, "ABCCED")); // true
        System.out.println("Search 'SEE': " + wordSearch(grid, "SEE"));       // true
        System.out.println("Search 'ABCB': " + wordSearch(grid, "ABCB"));     // false (can't reuse)
    }
}

File: Week 20/python/1.Backtracking.py
"""
WEEK 20 - PYTHON DSA
Topic: Backtracking — Template + Classic Problems
File: 1.Backtracking.py

CONCEPT:
    Backtracking incrementally builds a candidate solution. After every
    extension it checks whether the partial solution can still lead to a
    valid complete answer. If not, it ABANDONS that branch ("backtracks").

TEMPLATE:
    def backtrack(state):
        if is_solution(state):
            record(state); return
        for choice in choices(state):
            if is_valid(choice, state):
                make_choice(choice, state)   # explore
                backtrack(state)
                undo_choice(choice, state)   # backtrack

    The "undo" step distinguishes backtracking from plain DFS over a tree.

KEY POINTS:
    - Worst-case complexity is usually exponential (O(k^n) or O(n!)).
    - Pruning ("isValid") is the practical accelerator.
    - State is normally a mutable list/array — push then pop.

ALGORITHM / APPROACH:
    Four canonical problems (matching the Java reference):
        1. Permutations of a list of numbers
        2. Subsets (power set)
        3. N-Queens
        4. Word Search in grid

PYTHON-SPECIFIC NOTES:
    - Lists support efficient append/pop from the right -- the natural
      backtracking primitive.
    - Use a SHARED `current` list and snapshot it with `current[:]` or
      `list(current)` when adding to the result.
    - Recursion limit (default 1000) is plenty for these problems.
    - Yield from a generator is an alternative; we use plain accumulator
      lists for clarity.

DRY RUN:
    permutations([1,2,3])
        start=0
            swap(0,0) -> [1,2,3]; start=1
                swap(1,1) -> [1,2,3]; start=2
                    swap(2,2) -> emit [1,2,3]
                swap(1,2) -> [1,3,2]; start=2
                    swap(2,2) -> emit [1,3,2]
            (unswap back to [1,2,3])
            swap(0,1) -> [2,1,3]; ...
        eventually 6 permutations emitted.

    subsets([1,2,3])
        start=0 emit []
            include 1 -> [1] (start=1) emit [1]
                include 2 -> [1,2] emit; include 3 -> [1,2,3] emit; pop
                include 3 -> [1,3] emit; pop
            pop 1
            include 2 -> [2] emit; include 3 -> [2,3] emit; pop pop
            include 3 -> [3] emit
        Result has 8 = 2^3 subsets.

    nQueens(4) yields 2 solutions:
        .Q..      ..Q.
        ...Q      Q...
        Q...      ...Q
        ..Q.      .Q..

COMPLEXITY:
    permutations    O(n * n!) time, O(n) recursion depth
    subsets         O(n * 2^n) time, O(n) recursion depth
    nQueens         worst O(n!), pruned heavily in practice
    wordSearch      O(m*n * 4^len(word))
"""

from __future__ import annotations
from typing import List


# 1. PERMUTATIONS
def permutations(nums: List[int]) -> List[List[int]]:
    """All orderings of nums."""
    result: List[List[int]] = []

    def go(start: int) -> None:
        if start == len(nums):
            result.append(nums[:])         # snapshot
            return
        for i in range(start, len(nums)):
            nums[start], nums[i] = nums[i], nums[start]   # choose
            go(start + 1)                                 # explore
            nums[start], nums[i] = nums[i], nums[start]   # undo

    go(0)
    return result


# 2. SUBSETS (power set)
def subsets(nums: List[int]) -> List[List[int]]:
    """All 2^n subsets of nums."""
    result: List[List[int]] = []
    current: List[int] = []

    def go(idx: int) -> None:
        result.append(current[:])          # record current subset
        for i in range(idx, len(nums)):
            current.append(nums[i])        # include
            go(i + 1)
            current.pop()                  # backtrack

    go(0)
    return result


# 3. N-QUEENS
def n_queens(n: int) -> List[List[str]]:
    """All board configurations placing n non-attacking queens on n*n board."""
    result: List[List[str]] = []
    queens = [-1] * n                       # queens[row] = column

    def is_safe(row: int, col: int) -> bool:
        for r in range(row):
            c = queens[r]
            if c == col or abs(c - col) == abs(r - row):
                return False
        return True

    def build_board() -> List[str]:
        return ['.' * c + 'Q' + '.' * (n - c - 1) for c in queens]

    def go(row: int) -> None:
        if row == n:
            result.append(build_board())
            return
        for col in range(n):
            if is_safe(row, col):
                queens[row] = col
                go(row + 1)
                queens[row] = -1            # backtrack

    go(0)
    return result


# 4. WORD SEARCH IN GRID
def word_search(board: List[List[str]], word: str) -> bool:
    """Find word by traversing adjacent (no diagonal, no reuse) cells."""
    m, n = len(board), len(board[0])

    def dfs(i: int, j: int, k: int) -> bool:
        if k == len(word):
            return True
        if i < 0 or i >= m or j < 0 or j >= n:
            return False
        if board[i][j] != word[k]:
            return False
        tmp = board[i][j]
        board[i][j] = '#'                   # mark visited
        found = (dfs(i + 1, j, k + 1) or
                 dfs(i - 1, j, k + 1) or
                 dfs(i, j + 1, k + 1) or
                 dfs(i, j - 1, k + 1))
        board[i][j] = tmp                   # restore
        return found

    for i in range(m):
        for j in range(n):
            if dfs(i, j, 0):
                return True
    return False


def main() -> None:
    print("=== Permutations of [1,2,3] ===")
    for p in permutations([1, 2, 3]):
        print(p)

    print("\n=== Subsets of [1,2,3] ===")
    for s in subsets([1, 2, 3]):
        print(s)

    print("\n=== 4-Queens ===")
    sols = n_queens(4)
    print("Number of solutions:", len(sols))
    for sol in sols:
        for row in sol:
            print(row)
        print('---')

    print("\n=== Word Search ===")
    grid = [list("ABCE"), list("SFCS"), list("ADEE")]
    print("Search 'ABCCED':", word_search([r[:] for r in grid], "ABCCED"))  # True
    print("Search 'SEE'  :", word_search([r[:] for r in grid], "SEE"))      # True
    print("Search 'ABCB' :", word_search([r[:] for r in grid], "ABCB"))     # False


if __name__ == "__main__":
    main()


"""
NOTES (Python vs Java):
    - Java's char[][] mutation works for word search; Python uses list[list[str]]
      so we mutate cells likewise.
    - Snapshot mutable state with `current[:]` (slice) or `list(current)`
      before appending to results — equivalent to Java's `new ArrayList<>(current)`.
    - Tuple swap `a, b = b, a` is the natural in-place swap.
    - Recursion is idiomatic and adequate for these problem sizes; switch to
      iterative-stack only if recursion-depth becomes an issue.
    - N-queens row representation uses queens[row] = column; absolute-diff
      diagonal check is identical to the Java version.
"""
